functions/reacts.py

```diff
@@ -0,0 +1,156 @@
+    keywords: list[str]
+    # Whether the keywords should match the whole word or not.
+    match_whole_word: bool
+    # The link ID of the match. If the message matches the criteria, the link ID will be used to link the reaction/reply to the match.
+    match_link_id: str = ""
+
+# A response that can be triggered by a message that matches the criteria.
+class ReactResponse(BaseModel):
+    # The chance that the reaction/reply will be triggered. A float between 0 and 1.
+    chance: float
+    # The link ID of the match. If specified, the reaction/reply will only be triggered if the message matches the criteria and is linked to the match
+    match_link_id: str = ""
+    # A list of link IDs of other matches that the reaction/reply can be linked to.
+    other_match_link_ids: list[str] = []
+
+
+class ReactPossibleReaction(ReactResponse):
+    # A list of reactions that can be added to the message.
+    reactions: list[str]
+    # Whether all of the possible reactions should be added to the message.
+    react_with_all: bool = False
+    # The maximum number of reactions that can be added to the message. If -1, there is no limit.
+    max_react_limit: int = -1
+
+class ReactPossibleReply(ReactResponse):
+    # The reply message that can be sent.
+    message: str
+    # The type of the reply message.
+    type: ReactReplyType = "text"
+
+class ReactResource(BaseModel):
+    # A list of objects that define the criteria for a message to match the reaction. A message can match any of the criteria to trigger the possible responses.
+    criteria: list[ReactCriteria]
+    # A list of possible reactions that can be added to the message.
+    possible_reactions: list[ReactPossibleReaction] | None = None
+    # A list of possible replies that can be sent.
+    possible_replies: list[ReactPossibleReply] | None = None
+    
+
+REACT_RESOURCES_DIR = Path(Path(__file__).parent, "..", "resources", "reacts")
+
+
+def load_react_resources() -> dict[str, ReactResource]:
+    resources = {}
+    for file in REACT_RESOURCES_DIR.iterdir():
+        if file.suffix != ".json":
+            continue
+        with open(file, mode="r", encoding="utf8") as f:
+            filename = file.stem
+            data = json.load(f)
+            resources[filename] = ReactResource(**data)
+    return resources
+
+
+REACT_RESOURCES: dict[str, ReactResource] = load_react_resources()
+
+# For testing
+def check_resource_match(message_content: str, resource_name: str) -> bool:
+    resource = REACT_RESOURCES[resource_name]
+    matches = check_matches(message_content, resource.criteria)
+    return bool(matches)
+
+def check_matches(message_content: str, criteria: list[ReactCriteria]) -> bool | set[str]:
+    match_id_results = set()
+    found_match = False
+    for possible_match in criteria:
+        if possible_match.match_whole_word:
+            if compile(rf"\b(?:{'|'.join(possible_match.keywords)})\b", flags=IGNORECASE | UNICODE).search(message_content):
```

Comment:
I don't think `compile` is very efficient because it implies you're going to use it multiple times, usually with `compile` you define it as a global somewhere
should probably just use `re.match` or `search`